crates/sui-graphql-rpc/src/metrics.rs

```diff
@@ -1,72 +1,222 @@
+    pub(crate) fn get_query_path(&self, query: Vec<PathSegment>) -> String {
+        let mut path = String::new();
+        for (idx, s) in query.iter().enumerate() {
+            if idx > 0 {
+                path.push('.');
+            }
+            match s {
+                PathSegment::Index(idx) => {
+                    let _ = write!(&mut path, "{}", idx);
+                }
+                PathSegment::Field(name) => {
+                    let _ = write!(&mut path, "{}", name);
+                }
+            }
+        }
+        path
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct DBMetrics {
+    /// The number of fetches grouped by result (success or error)
+    pub db_fetches: IntCounterVec,
+    /// The fetch latency grouped by result (success or error)
+    pub db_fetch_latency: HistogramVec,
+    // TODO make this work, blocked by pg.rs (unclear if to use log function or smth else)
+    pub _db_query_cost: Histogram,
+    // TODO determine if we want this metric, and implement it
+    pub _db_fetch_batch_size: HistogramVec,
+}
+
+impl DBMetrics {
+    pub(crate) fn new(registry: &Registry) -> Self {
+        Self {
+            db_fetches: register_int_counter_vec_with_registry!(
+                "db_fetches",
+                "The number of fetches grouped by result (success or error)",
+                &["type"],
+                registry
+            )
+            .unwrap(),
+            db_fetch_latency: HistogramVec::new_in_registry(
+                "db_fetch_latency",
+                "The fetch latency grouped by result (success or error)",
+                &["type"],
+                registry,
+            ),
+            _db_query_cost: Histogram::new_in_registry(
+                "db_query_cost",
+                "Cost of a DB query",
+                registry,
+            ),
+            _db_fetch_batch_size: HistogramVec::new_in_registry(
+                "db_fetch_batch_size",
+                "Number of ids fetched per batch",
+                &["type"],
+                registry,
+            ),
+        }
+    }
 }
 
-// TODO: finetune buckets as we learn more about the distribution of queries
-const INPUT_NODES_BUCKETS: &[f64] = &[
-    1., 2., 4., 8., 12., 16., 24., 32., 48., 64., 96., 128., 256., 512., 1024.,
-];
-const OUTPUT_NODES_BUCKETS: &[f64] = &[
-    100., 200., 400., 800., 1200., 1600., 2400., 3200., 4800., 6400., 9600., 12800., 25600.,
-    51200., 102400.,
-];
-const QUERY_DEPTH_BUCKETS: &[f64] = &[
-    1., 2., 4., 8., 12., 16., 24., 32., 48., 64., 96., 128., 256., 512., 1024.,
-];
-const QUERY_PAYLOAD_SIZE_BUCKETS: &[f64] = &[
-    100., 200., 400., 800., 1200., 1600., 2400., 3200., 4800., 6400., 9600., 12800., 25600.,
-    51200., 102400.,
-];
-const DB_QUERY_COST_BUCKETS: &[f64] = &[
-    1., 2., 4., 8., 12., 16., 24., 32., 48., 64., 96., 128., 256., 512., 1024.,
-];
+#[derive(Clone)]
+pub(crate) struct RequestMetrics {
```

Comment:
We should also track the number of requests in flight -- this is usually done by counting the number of queries as they come into the system and as they leave the system.

Reply:
Sounds good!